bazel/internal.bzl

```diff
@@ -27,4 +27,10 @@ def redpanda_copts():
     copts.append("-Wno-missing-field-initializers")
     copts.append("-Wimplicit-fallthrough")
 
+    # for fmt v9 so that we do not need to write fmt::formatter wrappers for
+    # every output stream operator. note that we can't move to fmt >=v10 because
+    # this workaround macro was removed. so we'll need to rewrite the format
+    # handling for >1000 types.
+    copts.append("-DFMT_DEPRECATED_OSTREAM")
```

Comment:
Would we be able to write some magic template with sfinae magic to have a generic implementation? (which is what I'm assuming the macro is doing?)

Reply:
Ah I see. I think this will be a bunch of work and yes will require some clang refactoring tooling. I still think it will be a bunch of work tho :)

Reply:
yeh. very recently scylla just completed their conversion and final removal of this FMT_DEPRECATED_OSTREAM flag.

@tchaikov did you happen to use any tooling to automate some of that conversion, or just brute force?

Reply:
> which I'm guessing we can probably code generate pretty easy with some of those c++ parser tools you had mentioned?

It would be a little less work if you go through a layer of indirection.

I can't see a way to define a bunch of formatters without constantly going in and out of namespaces, which may be required if you want to format a bunch of nested types declared in the same namespace.

```c++
namespace n {

struct foo {
    // invalid code - can't specialise here
    // template<>
    // friend struct ::fmt::formatter<foo>;
};

// invalid code - can't specialise here
// template<>
// struct ::fmt::formatter<foo>;

} // close namespace

// error incomplete type in instantiation below
// template<>
// struct ::fmt::formatter<foo>;

// full specialisation must come before bar::operator<<
template<>
struct ::fmt::formatter<n::foo> : fmt::formatter<std::string_view> {
    template<typename FormatContext>
    auto format(const n::foo& b, FormatContext& ctx) const {
        return fmt::format_to(ctx.out(), "fmt::formatter<foo>");
    }
};

namespace n {

struct bar {
    foo _foo;
    friend std::ostream& operator<<(std::ostream& o, const bar& b) {
        fmt::print(o, "{}", b._foo); // fmt::formatter<foo> not found
        return o;
    }
};

} // n
```
If the type is a template, putting all the definitions in the implementation file is a pain even when all the instantiations are known.

Reply:
>// full specialisation must come before bar::operator<<

the godbolt provided in the fmt release notes put it after. does it matter?

https://godbolt.org/z/5sEc5qMsf

Reply:
> >// full specialisation must come before bar::operator<<
> 
> the godbolt provided in the fmt release notes put it after. does it matter?
> 
> https://godbolt.org/z/5sEc5qMsf

Yep, but no namespace, so you have to close, define the formatter, and reopen the namespace

Reply:
interesting. can you modify that godbolt link to demonstrate?

Reply:
> yeh. very recently scylla just completed their conversion and final removal of this FMT_DEPRECATED_OSTREAM flag.
> 
> @tchaikov did you happen to use any tooling to automate some of that conversion, or just brute force?

brute force. it was a 1-year effort for a single developer. see https://github.com/scylladb/scylladb/issues/13245

Reply:
> 1-year effort
@tchaikov wow!